### DISCUSSION TITLE
ScribeOcr Font Dection Algo

### DISCUSSION THREAD

**Discussion by 99bits** [scribeocr/scribeocr#52] [Q&A]
Thanks again for this awesome project.

One of the really cool feature of ScribeOcr is the ability  to replicate fonts, most OCRs are not doing this.  
We are trying to learn how this is done in ScribeOcr.  Can you share more detailed information where in the code is ScribeOcr is doing the font detection and the ideas behind it. 

Cheers

**Comment by ashokparmar29** [comment]
I am also interested in learning this. Please share information regarding font detection in Scribeocr.

**Comment by rmast** [comment]
As far as I understand ScribeOCR is just matching some default font groups and aligning spacing. It's not fully replicating the original fonts.

As I saw another solution, decapod, aiming at reconstructing the font, I investigated that. Decapod's font reconstruction only ended in it's early proof of concept stage. It also used fontforge for this True Type font reconstruction.
I myself think this font reconstruction should be combined with https://github.com/thuliu-yt16/dualvector by first reconstructing the font based on putting some images of the same glyph together at their center weight to be able to make use of their slight tilting on a line to find an even higher resolution of the original font than the scan resolution it was scanned with.

If those glyphs are for example matched after bicubic growing to about 512x512 or 1024x1024 and centering around their weight center before binarisation I guess they would make fine input for dualvector. The dualvector result could be tuned to their original glyphs.

I would not waste resolution for matching fonts on different sizes as Decapod did. Just match exactly the same glyphs that might have some subliminal subpixel shift that will be helpful for a better font reconstruction.

**Reply by 99bits** [reply]
Thanks @rmast  for your detailed message and analysis of the font matching challenge.  Out of curiosity, have you done some work on this?

**Comment by rmast** [comment]
The work I did led to my message.

I'm now two weeks off, so I try to concretize what I wrote in my spare time, I think the best spot for this work should be an extension to this repo.

Verzonden vanaf Outlook voor Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by 99bits** [reply]
Your message did sound like you dove quite deeply into this topic and that's why I asked.
Some concrete implementations would be amazing to build upon and extend upon this already awesome repos.

**Comment by rmast** [comment]
I did see DualVector is complete and runs without CUDA.

Verzonden vanaf Outlook voor Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by Balearica** [comment]
The response by @rmast is correct in that we are not either exactly matching the base font nor reconstructing it entirely from scratch.  A high-level description of the entire process is below.

Say that a document has text printed in the font "Helvetica Narrow Italic".
- The rough category of font is detected by the OCR engine.
   - We would know a font is "sans italic" based on data from the OCR engine.
   - This font information is accurately reported by Abbyy (for users who upload their own data) and the built-in Scribe OCR model, which is forked from Tesseract.
      - The standard Tesseract recognition model does not accurately report font data.
- A single font is chosen by comparing all available fonts within this category to the source image.
   - This narrows down the font from "Sans Italic" to "Nimbus Sans Italic" (Nimbus Sans is the closest font we have to Helvetica).
- Geometric transformations may be applied to the fonts to better match the character metrics reported by the OCR engine.
   - This allows the built in "Nimbus Sans" font to emulate Helvetica Narrow.
   - An example of font optimization can be seen in the readme [here](https://github.com/scribeocr/scribeocr?tab=readme-ov-file#proofreading-overview).

**Comment by Balearica** [comment]
Some miscellaneous notes regarding the existing approach, and other potential approaches, which may be useful to others looking into matching or replicating fonts--
- Our approach is heavily integrated into the OCR process.
   - There is no standalone font matching algorithm that could be used without running OCR.  Therefore, anybody looking for generic font identification/replication code to use with another project should probably look elsewhere.
- The approach of transforming a small number of fonts to replicate a large number of fonts was driven by experimentation--other approaches were attempted however were not as successful.
   - Attempting to match against a large font library is essentially impossible for legal reasons.
      - All fonts we use need to be permissively licensed (or at least GPL licensed), however there are many more font variants than you think, and most have no open source alternative. 
   - Attempting to generate entirely new fonts by vectorizing images requires did not work well using the tracing algorithms I tested.
      - To start, creating a new font by vectorizing images can only be attempted if the scan is very high-quality, which excludes most scans found in the wild.
         - Even screencaps--which most would consider to be "high quality"--generally are too low-resolution to even attempt tracing fonts.
      - Even with high-quality scans, the vectorization quality with commonly available algorithms was found to be poor.
         - I tested using `potrace` and `autotrace` as these are the algorithms used for the FontForge [autotrace](https://fontforge.org/docs/techref/autotrace.html) feature.  Neither worked well.
         - The core issue with these algorithms is that they very rarely produce clean, straight lines when they are supposed to.
            - What should be a clean line ends up being wavy; what should be crisp edges end up rounded off making the font look blob-y.
         - Having reviewed how these algorithms work, I think a new font-specific algorithm would be required to produce vectors with clean lines.  
      - I did not test with any other algorithms or ML-based models, so perhaps other solutions work better.
         - Anything added to the Scribe OCR site needs to run in the browser and add minimal size (<1MB), which eliminates most models that require large training data and a ML framework to run.
- The decision of what open source fonts to include is influenced by a custom font comparison developer tool--live at [fonts.scribeocr.com](https://fonts.scribeocr.com/), repo [here](https://github.com/scribeocr/font-comparison). 
   - This tool allows for comparing popular proprietary fonts to open source alternatives.
   - For example, selecting `Baskerville` and checking `Optimize` will show how well common open source alternative fonts Libre Baskerville and QTBasker replicate the most popular (proprietary) Baskerville font. 
   - Regarding the UI/output:
      - The `Optimize` checkbox enables the optimization feature seen in Scribe OCR, where fonts can be transformed to match the metrics in the source document.
      - The final tab of each comparison combines the best fitting glyphs across multiple fonts.
      - This site is for developer use, so it is not well documented and the interface is not pretty.  If anybody uses, you may need to review the source code to understand what is happening.

**Comment by 99bits** [comment]
Huge thanks @Balearica for the very detailed explanation of how all this works.  Your doing Font magic behind the scenes.
